### DISCUSSION TITLE
Question on visibility of image markers

### DISCUSSION THREAD

**Discussion by ovaisa** [iTwin/itwinjs-core#2204] [Q&A]
I have two questions on markers on which I would like to get advice.  Images below are taken from Design Review service and using Issues markers.  The first image shows the two issues makers I created to illustrate this problem.

1.  When section/clipping tool is used, the markers are not clipped.  Please see the second image below.  Model graphics are clipped but issues markers are still visible in the location where they were defined (even though that location in the model is no longer visible).

2.  Markers are visible even though their location is behind the model geometry, making it look like that they are at a location where they are not.  Pls see third image below.  The issue in red seems to be at nozzle whereas it is defined at a completely different location.

It seems GraphicsType (WorldOverlay ans Scene) doesn't seem to make difference.  It is probably that image graphics are drawn in view coordinates (rather than world) but for users this seems like error especially #2 as it makes these markers appear in wrong location.

Can you please guide how I can resolve these two problem.

![Marker-2](https://user-images.githubusercontent.com/83007711/132540730-a6214ca1-1aa9-4f6b-a35b-668573e900ea.JPG)

![Markers-1](https://user-images.githubusercontent.com/83007711/132540794-487512fc-9b52-4d37-b802-9e16cddfa1ce.JPG)

 
![Marker-3](https://user-images.githubusercontent.com/83007711/132540838-5e9b1927-ed9b-4d38-969f-b5b1e7145882.JPG)

**Comment by pmconne** [comment]
Show your code please. (As text, not screenshots, please).

If by "image marker" you mean [Marker](https://www.itwinjs.org/reference/imodeljs-frontend/views/marker/) then this is entirely expected - those are drawn as HTML canvas decorations, not WebGL graphics. Per the documentation:

> A Marker is a [[CanvasDecoration]], whose position follows a fixed location in world space.
> **Markers draw on top of all scene graphics**, and show visual cues about locations of interest.

**Reply by ovaisa** [reply]
Thanks for the clarification, certainly explains why these markers are appearing as they are.  I am afraid our users consider this an error as these seem to appear in the wrong location.  Hence I was seeking guidance on how we can resolve this.

I tried using 'graphical markers' where I used polyface geometry (with WorldDecoration option) instead of images by providing a draw function for the marker and that seems to work as users expect.  These markers get clipped by Clip tool and depth testing is enabled.  

However, we would like to use more visually appealing representation.  It is hard to code complex geometry using polyfaces + it is not customizable by users.  I wanted to know if there is way to use graphics similar to way MicroStation uses cells and cell libraries.  It is easy to build complex shapes in MicroStation as cells (and I assume they are also present in the synchronized imodel).  Is there a way to add/show this graphics as markers at locations of interest (similar to how cells are used in MicroStation).

This ability to show such graphics will also help with a construction planning capability we plan to work on in near future for users to place construction equipment such as trucks, cranes etc (using equipment models created by users) in an iModel.

Any guidance on how I can achieve this would be appreciated.

**Reply by pmconne** [reply]
What do you mean by "complex geometry using polyfaces"? Do you want your decorations to look just like the markers in your screenshots, always facing the camera, except with depth testing and perspective?

**Reply by ovaisa** [reply]
Ideally, yes!  We developed several SVG images (an example below) to depict different kind of locations of interest and ideally we  would like to show these graphics in the iModel at different locations depending on the situation.  We also developed replicas of these in MicroStation as cells but not sure how we can show them in iTwin viewer.


![icon](https://user-images.githubusercontent.com/83007711/132691555-eae7d132-3865-4941-8d2e-0edf908ba986.JPG)

**Reply by pmconne** [reply]
Are users supposed to be able to click on the markers to interact with them?

**Reply by ovaisa** [reply]
The interaction we are looking for is for the user to be able to click on the object to open more info about the location or issue.  The maker only shows the location and perhaps the type based on marker symbol.  Display of tooltip as the user hovers over the symbol is also very useful.

**Reply by pmconne** [reply]
You can do this with the existing GraphicBuilder APIs but it does require more code than a Marker.
- Create a RenderTexture from your image and a RenderMaterial from the texture.
- Create a GraphicBuilder, applying a transform that will rotate it to face the camera and specifying you don't want normals (no lighting).
- Activate the material.
- Use addShape to produce a rectangle.

A perhaps counter-intuitive but simpler approach would be to use a [particle collection](https://www.itwinjs.org/sample-showcase/?group=Viewer+Features&sample=fire-sample&imodel=Villa) for each marker, containing a single particle. Particles display an image on a rectangle that always faces the camera, and if defined in world coordinates they use depth testing.

In either case you would give each marker a unique Id (`IModelConnection.transientIds.next`) to make it clickable and have your Decorator handle the tooltip/click.

We can add to our backlog for 3.0 a simplified way of producing a billboard image to be drawn as a world decoration. We already have on that backlog a simplification of the RenderTexture and RenderMaterial APIs - they are the main complication with the GraphicBuilder approach (defining the geometry is quite easy). We also already have a way to make a graphic always face the camera, but there's no corresponding public API for it - we can add one.

**Reply by ovaisa** [reply]
Thank you Paul for you suggestions, I think the particle approach might be easier as I am not familiar with transformation required for billboarding.  If this code you mention to produce camera facing graphics is in the public domain part of iTwin, it would make it easier for me if I put that code in my vs project (or at least study those transformations).

More than billboard capability, what I feel would be more useful (at least for our project) is the ability to 'import' graphical objects (3d) and render them in iTwin.  This is similar to how Cesium allows rendering of external objects in glb/KML/Tile3d/etc formats.  We would like to bring in graphics created using 3d editors in iTwin and then render them where and when as needed.

https://sandcastle.cesium.com/?src=3D%20Models.html

MicroStation has cells and cell libraries (and now Component Center) to provide this capability.  Perhaps cells and cell libraries in a dgn are already in the iModel when the dgn is synchronized.  If the cell graphics is already in the iModel, is it possible to render this 'cells graphics' in iTwin.  This will allow us to create our markers as cells in MicroStation and place them in iTwin (hopefully with billboarding capability when available).  We also need this capability for our future workflows to let our users place construction equipment models in their digital twins (for construction planning).

**Reply by pmconne** [reply]
[Code that rotates particles to face camera](https://github.com/imodeljs/imodeljs/blob/a7f2ac51525898e1382f0eb3542c6d75cf831d66/core/frontend/src/render/ParticleCollectionBuilder.ts#L281).

Cesium supports glTF and glb, the standard web interchange formats for 3d objects; as well as 3d tile formats like b3dm, pnts, and so on, most of which use glTF under the hood. You can export a .obj to glb or glTF for viewing in Cesium.
iTwin.js supports displaying the same 3d tile formats, but doesn't currently directly support glTF and glb. Support for that should be trivial to add.

You can use TileAdmin.requestElementGraphics to obtain graphics for persistent elements, like in [this sample](https://www.itwinjs.org/sample-showcase/?group=Viewer+Features&sample=explode-sample).

**Reply by ovaisa** [reply]
Hi Paul,

From your response above (Sept 9, 2021) "We can add to our backlog for 3.0 a simplified way of producing a billboard image to be drawn as a world decoration.".

Wanted to check if the billboarding capability is now available.  If so, would be helpful if I can get an example or code snippet.

Thank you,

Ovais

**Comment by pmconne** [comment]
File an [issue](https://github.com/iTwin/itwinjs-core/issues/new?assignees=&labels=&template=feature_request.md&title=) describing in detail what you want.
